src/main/kotlin/org/rust/ide/intentions/ExtractEnumVariantIntention.kt

```diff
@@ -0,0 +1,210 @@
+
+    abstract fun replaceUsage(element: PsiElement, name: String)
+}
+
+private class TupleVariant(val fields: RsTupleFields, factory: RsPsiFactory) : VariantElement(fields, factory) {
+    override val visibility: RsVis? = extractVisibility(fields.tupleFieldDeclList.mapNotNull { it.vis }.toSet())
+    override val typeReferences: List<RsTypeReference> = fields.tupleFieldDeclList.map { it.typeReference }
+    override fun createStruct(name: String,
+                              typeParameters: String,
+                              whereClause: String): RsStructItem {
+        return factory.createStruct("struct $name$typeParameters${fields.text}$whereClause;")
+    }
+
+    override fun replaceUsage(element: PsiElement, name: String) {
+        val parent = element.ancestorStrict<RsPatTupleStruct>() ?: return
+        val binding = factory.createPatTupleStruct(name, parent.patList)
+
+        for (pat in parent.patList) {
+            val comma = pat.nextSibling
+            pat.delete()
+            if (comma.text == ",") {
+                comma.delete()
+            }
+        }
+        parent.addAfter(binding, parent.lparen)
+    }
+}
+
+private class StructVariant(val fields: RsBlockFields, factory: RsPsiFactory) : VariantElement(fields, factory) {
+    override val visibility: RsVis? = extractVisibility(fields.namedFieldDeclList.mapNotNull { it.vis }.toSet())
+    override val typeReferences: List<RsTypeReference> = fields.namedFieldDeclList.mapNotNull { it.typeReference }
+    override fun createStruct(name: String,
+                              typeParameters: String,
+                              whereClause: String): RsStructItem {
+        return factory.createStruct("struct $name$typeParameters$whereClause${fields.text}")
+    }
+
+    override fun replaceUsage(element: PsiElement, name: String) {
+        val parent = element.ancestorStrict<RsPatStruct>() ?: return
+        val path = parent.path
+        val binding = factory.createPatStruct(name, parent.patFieldList, parent.patRest)
+        val newPat = factory.createPatTupleStruct(path.text, listOf(binding))
+        parent.replace(newPat)
+    }
+}
+
+private fun extractVisibility(visibilities: Set<RsVis>): RsVis? {
+    return visibilities.maxBy {
+        when (val visibility = it.visibility) {
+            is RsVisibility.Public -> 4
+            is RsVisibility.Restricted -> if (visibility.inMod == it.crateRoot) {
+                3
+            } else {
+                2
+            }
+            is RsVisibility.Private -> 1
+        }
+    }
+}
+
+private data class HasTypeParameterVisitor(val parameter: RsTypeParameter) : TypeVisitor {
+    override fun visitTy(ty: Ty): Boolean = when (ty) {
+        is TyTypeParameter -> ty.name == parameter.name
+        else -> if (ty.needToSubstitute && ty.typeParameterValues[parameter] != null) {
+            true
+        } else {
+            ty.superVisitWith(this)
+        }
+    }
```

Comment:
As I know, `typeParameterValues` works correctly only for ADT's. So you need something like that:
```suggestion
    override fun visitTy(ty: Ty): Boolean =
        when {
            ty is TyTypeParameter -> ty.name == parameter.name
            ty.hasTyTypeParameters -> ty.superVisitWith(this)
            else -> false
        }
```